New functionability in pandas_options using 'dtype':str

cjotade opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
Changes has been made in _extraxt_from method from io module allowing input in pandas_options the 'dtype'=str. Note that before these fixes you use pd.to_numeric in df.columns to force the dataframe be numeric, which didn't allow return the frame as a string. For instance, if 
pandas_options['dtype'] == str then don't force the numeric frame and parse the 'None' string values to None. See Figure 4.
## Motivation and Context
<!--- Why is this change required? What problem does it solve? --> This fix gives the possibility to return the frame with string values without numeric issues. For example, if you have some data which uses European decimals like 1.321,21 and 112,21 and 1.421 then with pd.to_numeric the first two numbers return will be '1.321,21' and '112,21' because pandas couldn't force numeric. On the other hand, the return 1.421 in pd.to_numeric will be 1.421 but it should be 1421. The problem here is that pd.DataFrame don't have the parameter decimals unlike pd.read_csv.

<!--- If it fixes an open issue, please link to the issue here. -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. --> Tested using read_pdf_with_template with many crops. 
<!--- Include details of your testing environment, tests ran to see how --> In Figure 1 the multiplications of columns Quant with Valor unit (R$) must be the column Valor (R$), so I created a pytest environment to test if this operation works in many pdf's, with my fixes the tests passed for 24 tables.
<!--- your change affects other areas of the code, etc. --> My changes only affects when you're using multiple_tables=True (or read_pdf_with_template with multiple tables) and when the pandas_options 'dtype' key is str.

## Screenshots (if appropriate):
In the image, the Quant column 17.880 should be 17880 but with the pd.to_numeric it returns 17.88, see Figure 2 and compare with the fix in Figure 3: 
![Screenshot from 2020-02-20 18-32-56](https://user-images.githubusercontent.com/43217827/74980966-2d091000-5410-11ea-8f4f-46f09cf6e4b2.png)
Figure 1. Table in pdf.

Return without the fixes (using df.to_json() just for seing the dtypes):
![Screenshot from 2020-02-20 19-01-15](https://user-images.githubusercontent.com/43217827/74982712-68590e00-5413-11ea-91b4-c04add57afd0.png)
Figure 2. Returned df.to_json() without fixes.

Return with fixes (using df.to_json() just for seing the dtypes):
![Screenshot from 2020-02-20 18-58-20](https://user-images.githubusercontent.com/43217827/74982557-1adca100-5413-11ea-9d96-b9e3de2a7c10.png)
Figure 3. Returned df.to_json() with fixes.

In the follow image the changes which were made in the _extraxt_from method from io module:
![Screenshot from 2020-02-20 18-43-57](https://user-images.githubusercontent.com/43217827/74981418-fd0e3c80-5410-11ea-9edb-3ab029c53b57.png)
Figure 4. Changes in code.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I read [the contributing document](https://tabula-py.readthedocs.io/en/latest/contributing.html).
- [x] My code follows the code style of this project with running linter.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

chezou commented:
@cjotade thanks you for your contribution. Motivation should be reasonable. But, this implementation doesn't care about other dtypes like int, or dict style option.
https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html

Could you consider to accept all the possible options for dtype?

cjotade commented:
Thanks for your suggestion @chezou. So I have a question about the dict style option, in our approach we have many tables or crops, let say N tables ![image](https://user-images.githubusercontent.com/43217827/75058725-4a92b400-54ba-11ea-8d6c-5b482091fcca.png) 
 and every df has shape ![image](https://user-images.githubusercontent.com/43217827/75059006-d1479100-54ba-11ea-85f1-7b70a959ff46.png), hence How we know the specific shape for every crop? cause the number of columns in every table is variable, maybe the input parameter dtype should be a nested dict like the templates?

chezou commented:
good point. I meant pandas_option should align with original option as much possible because tabula-py users assume that option is compatible with pandas.

Now I understand why you implemented str only. I think it'd be better to treat supported dtypes not only str except for dict, or at least we should raise exception with unexpected options. How do you feel?

chezou commented:
btw, pandas read_csv has thousands and decimal options. Do those options work with your data?

cjotade commented:
you're right, I think that the option is handle dict dtype with exceptions, for the other dtypes I'll align them as the pandas original option and text you back. 

read_csv actually has the decimal option, but in my case I had two problems. The first, my crops have standard and European decimals mixed, so for my problem was better transform them to str and then fix every crop separately. The second problem is that despite read_csv has decimal option, pd.DataFrame does't, so if you have many crops, then the module io.py will use just pd.DataFrame instead of read_csv, so it is not possible to solve the problem.

cjotade commented:
I notice that pd.DataFrame raises an exception for dict dtype, so its not necessary re implement it. So I just extends the code below for int. Do we need any other dtype? 

![Screenshot from 2020-02-27 16-59-01](https://user-images.githubusercontent.com/43217827/75481869-7f06e400-5982-11ea-83b4-3f43d0829c8d.png)

chezou commented on tabula/io.py:
```diff
@@ -678,10 +678,19 @@ def _extract_from(raw_json, pandas_options=None):
 
-        for c in df.columns:
-            df[c] = pd.to_numeric(df[c], errors="ignore")
+        if pandas_options.get('dtype', None) == int:
```
I misunderstood your intention.
I think this None value handling for str is redundant because `dtype=str` keeps null value as `np.nan` without modification.
Just adding

```
if not pandas_options.get("dtype"):
```

before `for c in df.columns` would be enough.

Could you check it?

chezou commented on tabula/io.py:
```diff
@@ -678,10 +678,19 @@ def _extract_from(raw_json, pandas_options=None):
+                df.loc[mask[col], col] = None
+        else:
+            for c in df.columns:
+                df[c] = pd.to_numeric(df[c], errors="ignore")
```
I mean, this would be enough for your motivation. Could you try it with your data?

```suggestion
        if not pandas_options.get('dtype'):
            for c in df.columns:
                df[c] = pd.to_numeric(df[c], errors="ignore")
```

cjotade replied:
Its works in my data with your suggestion :100:

chezou commented on tabula/io.py:
```diff
@@ -678,10 +678,19 @@ def _extract_from(raw_json, pandas_options=None):
+                df[c] = pd.to_numeric(df[c], errors="ignore")
         data_frames.append(df)
 
+
```
Needless extra line

```suggestion
```

chezou replied:
Could you remove this line?

cjotade commented:
I added a commit with your suggestion, which works!

chezou commented:
Thanks for fixing. Could you add unit test as well? If needed, it'd be great if you could add some PDF for test as well.

cjotade commented:
I added a new test method called test_read_pdf_with_dtype_string, also added one new pdf which is the same I showed you in the Figure 1. The method tests also with your pdf's

cjotade commented on tests/test_read_pdf_table.py:
```diff
@@ -292,6 +292,25 @@ def test_read_pdf_with_jar_path(self, jar_func, mock_fun):
+
+        pandas_options = {'dtype': str}
+        df = tabula.read_pdf(self.pdf_path, stream=True, pages=0, pandas_options=pandas_options.copy())
+        self.assertTrue(df.equals(pd.read_csv(self.expected_csv1, **pandas_options.copy())))
```
Its seems if in some page there is just one table, read_pdf return one dataframe and not a list of dfs. In the code above if you set pages=2 the assert breaks cause in these page there are more than one table.

cjotade commented:
Adding changes that you said, I keep an eye to your comments

chezou commented:
hmm, test on master branch failed. Will take a look.

cjotade commented:
Many Thanks!

chezou commented:
@cjotade Released on PyPI v2.1.0 https://pypi.org/project/tabula-py/2.1.0/
